CONTRIBUTING.md

```diff
@@ -0,0 +1,114 @@
+
+* Compiler tests -- these tests make sure that olive language compiler produce the expected output (either success or error messages)
+  * Run with `mvn clean install`
+  * Tests are in `shesmu-server/src/test/resources/compiler` and `shesmu-server/src/test/resources/compiler`
+  * All new syntax changes require tests
+  * Fixes for bad bytecode generation require tests
+* User interface tests
+  * Run with `mvn clean install`
+  * Tests are in `shesmu-server/src/test/java/ca/on/oicr/gsi/UserInterfaceTest.java`
+* Miscellaneous tests
+  * Run with `mvn clean install`
+  * Tests are in `shesmu-server/src/test/java/ca/on/oicr/gsi/`
+* Plugin tests
+  * Run with `mvn clean install`
+
+Please make sure to add or update the relevant tests for your changes. Testing
+plugins is difficult because they are often mostly integration code that has to
+communicate with a external services. We accept any level of testing on plugins
+as long as it is fast and delegate any plugin bugs to plugin authors.
+
+## Commits
+
+* Make sure your commit contains a reference to the issue number. _e.g._
+  `Closes #1234` in the body Edit the **Unreleased** section in
+* `RELEASE_NOTES.md` to detail any user-visible **Changes** or
+	**Update Notes** (Additional steps that must be taken when upgrading to the
+  Shesmu version containing your change)
+
+## Pull Requests
+
+Changes should never be merged directly into `master`. Pull requests should be
+made into the `master` branch for testing and review. All pull requests need
+two reviewers. If you have suggestions, please select them when creating your
+pull request, but the Shesmu developers may add additional developers or assign
+someone else.
+
+## Merging
+
+Once all of the tests are passing, and your pull request has received two
+approvals, you are ready to merge. To keep a clean commit history please
+
+1. Merge your changes into one commit unless they are clearly separate changes.
+2. Rebase on the `master` branch so that your change appears after the changes
+   that were previously merged in the history.
+
+To do this, on your system:
+
+    git fetch && git rebase -i origin/master
+
+Then change any commits that have no semantic meaning to `f` fixup commits to
+be absorbed into the previous commit. Finally, push the modified version to
+your branch with:
+
+    git push -f
+
+Please delete your feature branch after it is merged.
+
+## Plugins versus Core Infrastructure
+Shesmu has two very different part: the core Shesmu infrastructure (the server,
+the compiler, and testing infrastructure) and plugins for integration with
+other systems. We are happy to take on plugins for systems we do not use and
+provide them with some maintenance and to include them in our release process
+to make them as available as possible.
+
+However, the core Shesmu team cannot have adequate experience and testing for
+every plugin and the software they use. If you are contributing a plugin, there
```

Comment:
Grammar is funny. How about "the core Shesmu team does not have the testing environment and therefore cannot support every plugin contributed by the community".